sys/include/net/sock/udp.h

```diff
@@ -352,6 +352,14 @@ typedef struct {
     sock_aux_flags_t flags; /**< Flags used request information */
 } sock_udp_aux_tx_t;
 
+/**
+ * @brief   A data chunk
+ */
+typedef struct {
+    const void *data;   /**< payload      */
+    size_t len;         /**< payload size */
+} sock_udp_data_snip_t;
```

Comment:
Why not use https://api.riot-os.org/iolist_8h.html ?

This is also what is presented at very bottom of the stack to the netdev, so using it here could make some nice optimisations possible. Also, since that type is bound to be used by drivers anyway, reusing the helper functions if that type could safe some rom of the same functionality is needed by both driver and application.

Reply:
This would make `sock` unusable outside the RIOT-context, though, which used to be a design goal (not sure if with the most recent changes this is still the case though). One should be able to use `sock` on Linux as well (this was how nanocoap was tested back then).

Reply:
Hm `iolist_t` implies a linked list which makes this a bit more cumbersome to use.
It is also translated into `gnrc_pktsnip_t` down the stack, so we don't benefit from type reuse.

Reply:
> This would make `sock` unusable outside the RIOT-context

Do you mean with `iolist` or with the new API? Last time I checked there is no `sock_udp_send_aux()` on Linux, so we'd need a wrapper either way.

Reply:
Does iol contain anything that would be difficult to provide outside of RIOT?

Also, isn't that use case covered already by the `native` board?

Reply:
> Hm `iolist_t` implies a linked list which makes this a bit more cumbersome to use. It is also translated into `gnrc_pktsnip_t` down the stack, so we don't benefit from type reuse.

Couldn't you just enque them and adapt the free function to stop the iteration after the transport layer snip? The pktsnips have iolist as "base class" structure anyway.

Reply:
> The pktsnips have iolist as "base class" structure anyway.

No [they don't](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/pkt.h#L108).

Reply:
It would turn enquing them into O(1), but given that it is unlikely that there are more than two application layer fragments, it seems not to be worth the effort.